src/Compilers/CSharp/Portable/Utilities/ValueSetFactory.BoolValueSet.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+
+#nullable enable
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    using static BinaryOperatorKind;
+
+    internal static partial class ValueSetFactory
+    {
+        private class BoolValueSet : IValueSet<bool>
+        {
+            private readonly bool _hasFalse, _hasTrue;
+
+            internal static BoolValueSet AllValues = new BoolValueSet(hasFalse: true, hasTrue: true);
+            internal static BoolValueSet None = new BoolValueSet(hasFalse: false, hasTrue: false);
+            internal static BoolValueSet OnlyTrue = new BoolValueSet(hasFalse: false, hasTrue: true);
+            internal static BoolValueSet OnlyFalse = new BoolValueSet(hasFalse: true, hasTrue: false);
+
+            private BoolValueSet(bool hasFalse, bool hasTrue) => (_hasFalse, _hasTrue) = (hasFalse, hasTrue);
+
+            public static BoolValueSet Create(bool hasFalse, bool hasTrue)
+            {
+                switch (hasFalse, hasTrue)
+                {
+                    case (false, false):
+                        return None;
+                    case (false, true):
+                        return OnlyTrue;
+                    case (true, false):
+                        return OnlyFalse;
+                    case (true, true):
+                        return AllValues;
+                }
+            }
+
+            bool IValueSet.IsEmpty => !_hasFalse && !_hasTrue;
+
+            IValueSetFactory<bool> IValueSet<bool>.Factory => BoolValueSetFactory.Instance;
+
+            IValueSetFactory IValueSet.Factory => BoolValueSetFactory.Instance;
+
+            public bool Any(BinaryOperatorKind relation, bool value)
+            {
+                switch (relation, value)
+                {
+                    case (Equal, true):
+                        return _hasTrue;
+                    case (Equal, false):
+                        return _hasFalse;
+                    default:
+                        throw new ArgumentException("relation");
+                }
+            }
+
+            bool IValueSet.Any(BinaryOperatorKind relation, ConstantValue value) => value.IsBad || Any(relation, value.BooleanValue);
```

Comment:
>value.IsBad || [](start = 84, length = 14)

Why isn't this checked in the `Any()` implementation above?

More generally, could the `IValueSet` implementation simply delegate to the `IValueSet<T>` implementation, here and in other `ValueSet` classes?